libcontainer/nsenter/nsexec.c

```diff
@@ -632,6 +638,76 @@ void join_namespaces(char *nslist)
 	free(namespaces);
 }
 
+/*
+ * enable_ima_ns - sets up IMA Namespace for the process
+ *
+ * In order to set it up correctly calling process
+ * should have new userns first, because if it is in the first userns then
+ * new imans won't be created (imans has no CLONE_* flags). The function does next things:
+ *
+ * 1. It mounts temporary securityfs, in order to not bother hosts's
+ * filesystem, we've chosen container's rootfs as a location for this,
+ * (to be more accurate we've chosen /mnt inside the rootfs). Make
+ * sure the calling process already created new userns.
+ *
+ * 2. It writes '1' to rootfs/mnt/integrity/ima/active. This is
+ * the key stage of enabling imans. After that imans will be actually created.
+ *
+ * 3. Unmount temporary securityfs.
+ *
+ */
+#define IMA_ACTIVE_PATH	"integrity/ima/active"
+/* IMA  securityfs relative mount point */
+#define IMA_SECURITYFS_REL_MNT	"/mnt/"
+
+static void enable_ima_ns(void)
+{
+	int err, mntflags, ima_active_path_len;
+	char data, ima_active[PATH_MAX + 1];
+
+	ima_active_path_len = strlen(IMA_ACTIVE_PATH);
+	/*
+	 * Now current working directory is our rootfs,
+	 * so in order to avoid some conflicts with Host filesystem,
+	 * we should mount securityfs inside the rootfs of the currently
+	 * starting container. The mount point for this securityfs is
+	 * CWD + "/mnt"
+	 */
+	if(!getcwd(ima_active, PATH_MAX))
+		bail("failed to get path of the current rootfs");
+
+	if ((strnlen(ima_active, PATH_MAX) + strlen(IMA_SECURITYFS_REL_MNT)) > PATH_MAX)
+		bail("securityfs temporary path is too long");
+
+	strncat(ima_active, IMA_SECURITYFS_REL_MNT, PATH_MAX);
+	mntflags = MS_NOATIME | MS_NODEV | MS_NOEXEC | MS_NOSUID | MS_RELATIME;
+	err = mount("securityfs", ima_active, "securityfs", mntflags , NULL);
+	if (err != 0)
+		bail("failed to mount temporary securityfs");
+
+	/*
+	 * According to security/integrity/ima/ima_fs.c:ima_write_active()
+	 * the length of the buffer <= 2 and it is only allowed to write '1', '1\0' or '1\n'
+	 */
+	if ((strnlen(ima_active, PATH_MAX) + ima_active_path_len) > PATH_MAX)
+		bail("ima active path is too long");
```

Comment:
Perhaps move this check to before doing `mount()`? Also, since we can assume (and do a run-time assert in the code) that `ima_active_path_len > strlen(IMA_SECURITYFS_REL_MNT)`, we can omit the first length check.

Reply:
I don't think that we can completely omit the first check `len(ima_active) + len(IMA_SECURITYFS_REL_MNT) > PATH_MAX`, replacing it with `len(ima_active) + ima_active_path_len > PATH_MAX`, because before `write_file` we'll do `strncat(ima_active, IMA_ACTIVE_PATH)`, where `ima_active` includes `IMA_SECURITYFS_REL_MNT`.

But you're right it is possible for the second check to do it before `mount` like this:
```
if ((len(ima_active) + len(IMA_SECURITYFS_REL_MNT) + ima_active_path_len) > PATH_MAX)
    bail(...)
```
This will tell us whether `len('/path/to/rootfs') + len('/mnt/') + len('integrity/ima/active')` is more than `PATH_MAX`. So, no need to do any length check after `mount`.

Thanks!